src/go/serviceconfig/service_config_fetcher.go

```diff
@@ -61,37 +60,61 @@ func NewServiceConfigFetcher(mf *metadata.MetadataFetcher, opts options.ConfigGe
 			Timeout: opts.HttpRequestTimeout,
 		},
 		serviceName: serviceName,
-		mf:          mf,
 		opts:        opts,
 	}
+
+	configIdFetcher := NewConfigIdFetcher(serviceName, opts.ServiceControlURL,
+		scf.client, func() (string, time.Duration, error) { return scf.accessToken() })
+	scf.newConfigId = func() (string, error) {
+		return configIdFetcher.fetchNewConfigId()
+	}
+
+	scf.accessToken = func() (string, time.Duration, error) {
+		if accessTokenFromImds == nil && scf.opts.ServiceAccountKey == "" {
+			return "", 0, fmt.Errorf("If --non_gcp is specified, --service_account_key has to be specified.")
+		}
+		if scf.opts.ServiceAccountKey != "" {
+			return util.GenerateAccessTokenFromFile(scf.opts.ServiceAccountKey)
+		}
+		return accessTokenFromImds()
```

Comment:
I don't understand: Why does this call `accessTokenFromImds()`, but in this line below you call `mf.FetchAccessToken()`.

https://github.com/GoogleCloudPlatform/esp-v2/pull/83/files#diff-f2a296fe272e72d0390a8830217a156aR141

Shouldn't both of these be using the same access token?

Additionally, this looks like you fetch the access token every time you make a call? Could you clarify how the access token is fetched / cached?

Reply:
- It is just a wrapper of mf.FetchAccessToken()
- It is cached in mf.FetchAccessToken has cache mechanism but I admit the name looks like just simply fetching.